src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,192 @@
+//! Paths and Unix shells
+//!
+//! MacOS, Linux, FreeBSD, and many other OS model their design on Unix,
+//! so handling them is relatively consistent. But only relatively.
+//! POSIX postdates Unix by 20 years, and each "Unix-like" shell develops
+//! unique quirks over time.
+//!
+//!
+//! Windowing Managers, Desktop Environments, GUI Terminals, and PATHs
+//!
+//! Duplicating paths in PATH can cause performance issues when the OS searches
+//! the same place multiple times. Traditionally, Unix configurations have
+//! resolved this by setting up PATHs in the shell's login profile.
+//!
+//! This has its own issues. Login profiles are only intended to run once, but
+//! changing the PATH is common enough that people may run it twice. Desktop
+//! environments often choose to NOT start login shells in GUI terminals. Thus,
+//! a trend has emerged to place PATH updates in other run-commands (rc) files,
+//! leaving Rustup with few assumptions to build on for fulfilling its promise
+//! to set up PATH appropriately.
+//!
+//! Rustup addresses this by:
+//! 1) using a shell script that updates PATH if the path is not in PATH
+//! 2) sourcing this script in any known and appropriate rc file
+
+use super::canonical_cargo_home;
```

Comment:
canonical_cargo_home is a bit of an odd function - it transforms any non-clean unicode characters in a bid to make the path printable. This matters when e.g. folk have paths on disk that are in SHIFT-JIS but their encoding is in CP1252 or other incompatible combinations.

So rather than using this function designed for display, call `cargo_home()?.join("bin").to_str()`. Factoring out a helper for that is *just about* a common enough thing to be worth doing, but I haven't gotten around to it yet.

If `to_str()` fails, then we should fail to write the env file for now (as it would be written with bogus content). A more sophisticated iteration of this work would be to write the env file as a bytestring, not unicode, so that we can write it even if cargo_home isn't valid unicode. I don't know if you want to tackle that right now or not.

Reply:
The function should Err in that case, right? What Err should it return? I'm struggling a bit to interpret how to make err_chain work here, looking at it. to_str() gives an Option which I can .ok_or() on. 

This dilemma crops up on either path. Frankly it seems like I'm going to go to the same amount of trouble either way so I might as well write some raw bytes, but we'll see.

Reply:
if you choose to defer the work and return an error, you'd say something like

```
let as_str = if let Some(s) = p.to_str() { s } else bail!("non-unicode homedir")
```

Reply:
Ah, that is simpler than expected then, so I will do that and come back to write the bytestring.

Reply:
A later patch, however.

I should note that Rustup was previously using the [lossy canonical_cargo_home]https://github.com/rust-lang/rustup/blob/7050161b9d89afd1db7d3d374b1acc2bb368383e/src/cli/self_update/unix.rs#L95-L100) in this segment of the code anyways so this is already doing additional fixes.